lib/src/chunked_stream_reader.dart

```diff
@@ -0,0 +1,176 @@
+    return result;
+  }
+
+  /// Read next [size] elements from _chunked stream_ as a sub-stream.
+  ///
+  /// This will pass-through _chunks_ from the underlying _chunked stream_ until
+  /// [size] elements have been returned, or end-of-stream has been encountered.
+  ///
+  /// If end-of-stream is encountered before [size] bytes is read, this returns
+  /// a _chunked stream_ with less than [size] (indicating end-of-stream).
+  ///
+  /// If the underlying stream throws, the stream is cancelled, the exception is
+  /// propogated and further read operations will fail.
+  ///
+  /// If the sub-stream returned from [readStream] is cancelled the remaining
+  /// unread elements up-to [size] are drained, allowing subsequent
+  /// read-operations to proceed after cancellation.
+  ///
+  /// Throws, if another read-operation is on-going.
+  Stream<List<T>> readStream(int size) {
+    if (size < 0) {
+      throw ArgumentError.value(size, 'size', 'must be non-negative');
+    }
+    if (_reading) {
+      throw StateError('Concurrent read operations are not allowed!');
+    }
+    _reading = true;
+
+    final c = StreamController<List<T>>();
+
+    c.addStream(() async* {
+      // While we have data to read
+      while (size > 0) {
+        // Read something into the buffer, if it's empty
+        if (_buffer.isEmpty) {
+          if (!(await _input.moveNext())) {
+            // Don't attempt to read more data, as there is no more data.
+            size = 0;
+            _reading = false;
+            break;
+          }
+          _buffer = _input.current;
+        }
+
+        if (size < _buffer.length) {
+          final output = _buffer.sublist(0, size);
+          _buffer = _buffer.sublist(size);
+          size = 0;
+          yield output;
+          _reading = false;
+          break;
+        }
+
+        if (_buffer.isNotEmpty) {
+          final output = _buffer;
+          size -= _buffer.length;
+          _buffer = _emptyList;
+          yield output;
+        }
+      }
+    }()).whenComplete(c.close);
+
+    c.onCancel = () async {
+      while (size > 0) {
+        if (_buffer.isEmpty) {
+          if (!await _input.moveNext()) {
+            size = 0; // no more data
+            _reading = false;
+            break;
+          }
+          _buffer = _input.current;
+        }
+
+        if (size < _buffer.length) {
+          _buffer = _buffer.sublist(size);
+          size = 0;
+          _reading = false;
+          break;
+        }
+
+        size -= _buffer.length;
+        _buffer = _emptyList;
+      }
+    };
+
+    return c.stream;
+  }
+
+  /// Cancel the underlying _chunked stream_.
+  ///
+  /// It is always safe to call [cancel], even if the undelying stream was read
+  /// to completion.
+  ///
+  /// It can be a good idea to call [cancel] in a `finally`-block when done
+  /// using the [ChunkedStreamReader], this mitigates risk of leaking resources.
+  Future<void> cancel() async => await _input.cancel();
```

Comment:
Should there be a state check here? No cancelling while a read is in progress, otherwise ... well, either you get a state error from the stream iterator immediately, or if you do this at a yield-point, the next `moveNext` will return false. That seems safe, but possibly surprising that the input is cut short from the side.

Reply:
From reading the `StreamIterator` code I get the impression that `StreamIterator.cancel` makes any pending future from `StreamIterator.moveNext` return `false`.

So if `ChunkedStreamReader.cancel` is called while a future `readChunk` is pending, we get the same behavior as if the underlying stream ended early. Namely `readChunk(size)` may return a chunk smaller than the requested `size`.

I added a test to cover both cases...

Reply:
You can cancel while moving! Yey! Then LGTM.